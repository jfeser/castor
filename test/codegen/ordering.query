alist(orderby([f desc], r1) as k, atuple([ascalar(k.f), ascalar(k.g)], cross))