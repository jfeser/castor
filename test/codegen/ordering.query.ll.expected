; ModuleID = 'scanner'
source_filename = "scanner"

%params = type { [1 x i64]* }

@true_str = private unnamed_addr constant [2 x i8] c"t\00", align 1
@false_str = private unnamed_addr constant [2 x i8] c"f\00", align 1
@null_str = private unnamed_addr constant [5 x i8] c"null\00", align 1
@void_str = private unnamed_addr constant [3 x i8] c"()\00", align 1
@sep_str = private unnamed_addr constant [2 x i8] c"|\00", align 1
@newline_str = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@int_fmt = private unnamed_addr constant [3 x i8] c"%d\00", align 1
@str_fmt = private unnamed_addr constant [5 x i8] c"%.*s\00", align 1
@float_fmt = private unnamed_addr constant [3 x i8] c"%f\00", align 1
@date_fmt = private unnamed_addr constant [15 x i8] c"%04d-%02d-%02d\00", align 1

; Function Attrs: argmemonly nofree nosync nounwind willreturn
declare void @llvm.lifetime.start.p0i8(i64 immarg, i8* nocapture) #0

; Function Attrs: argmemonly nofree nosync nounwind willreturn
declare void @llvm.lifetime.end.p0i8(i64 immarg, i8* nocapture) #0

; Function Attrs: argmemonly readonly speculatable
declare i32 @cmph_search_packed(i8*, i8*, i32) #1

declare i32 @printf(i8*, ...)

; Function Attrs: argmemonly norecurse nounwind readonly speculatable
declare i32 @strncmp(i8* noalias nocapture, i8* noalias nocapture, i64) #2

declare i8* @strncpy(i8*, i8*, i32)

; Function Attrs: argmemonly norecurse nounwind readonly speculatable
declare i64 @strpos(i8* nocapture, i64, i8* nocapture, i64) #2

; Function Attrs: readnone
declare i64 @extract_year(i64) #3

; Function Attrs: readnone
declare i64 @extract_month(i64) #3

; Function Attrs: readnone
declare i64 @extract_day(i64) #3

; Function Attrs: readnone
declare i64 @add_month(i64, i64) #3

; Function Attrs: readnone
declare i64 @add_year(i64, i64) #3

; Function Attrs: argmemonly norecurse nounwind readonly speculatable
define i1 @scmp({ i8*, i64 } %0, { i8*, i64 } %1) #2 {
entry:
  %len = extractvalue { i8*, i64 } %0, 1
  %pos = extractvalue { i8*, i64 } %0, 0
  %len1 = extractvalue { i8*, i64 } %1, 1
  %pos2 = extractvalue { i8*, i64 } %1, 0
  %len_cmp = icmp eq i64 %len, %len1
  br i1 %len_cmp, label %eq, label %neq

eq:                                               ; preds = %entry
  %str_cmp = call i32 @strncmp(i8* %pos, i8* %pos2, i64 %len), !tbaa !0
  %str_cmp3 = icmp eq i32 %str_cmp, 0
  ret i1 %str_cmp3

neq:                                              ; preds = %entry
  ret i1 false
}

; Function Attrs: argmemonly norecurse nounwind readonly
define void @printer(%params* noalias %0) #4 {
entry:
  %tupleptrtmp = alloca { i64, i64 }, align 8
  %i1 = alloca i64, align 8
  %cstart4 = alloca i64, align 8
  %cstart3 = alloca i64, align 8
  %cstart0 = alloca i64, align 8
  %count2 = alloca i64, align 8
  store i64 0, i64* %cstart0, align 8, !tbaa !3
  store i64 0, i64* %i1, align 8, !tbaa !3
  store i64 5, i64* %count2, align 8, !tbaa !3
  br label %loopcond

loopcond:                                         ; preds = %entry
  %i11 = load i64, i64* %i1, align 8, !tbaa !3
  %count22 = load i64, i64* %count2, align 8, !tbaa !3
  %lttmp = icmp slt i64 %i11, %count22
  br i1 %lttmp, label %loopbody, label %loopend

loopbody:                                         ; preds = %loopbody, %loopcond
  %cstart03 = load i64, i64* %cstart0, align 8, !tbaa !3
  store i64 %cstart03, i64* %cstart3, align 8, !tbaa !3
  %cstart34 = load i64, i64* %cstart3, align 8, !tbaa !3
  %addtmp = add nsw i64 %cstart34, 1
  store i64 %addtmp, i64* %cstart4, align 8, !tbaa !3
  %cstart35 = load i64, i64* %cstart3, align 8, !tbaa !3
  %buf = getelementptr inbounds %params, %params* %0, i32 0, i32 0
  %buf_ptr = load [1 x i64]*, [1 x i64]** %buf, align 8, !tbaa !5
  %buf_ptr_cast = bitcast [1 x i64]* %buf_ptr to i8*
  %slice_ptr = getelementptr inbounds i8, i8* %buf_ptr_cast, i64 %cstart35
  %slice_val = load i8, i8* %slice_ptr, align 1, !tbaa !7
  %int_val = sext i8 %slice_val to i64
  %cstart46 = load i64, i64* %cstart4, align 8, !tbaa !3
  %buf7 = getelementptr inbounds %params, %params* %0, i32 0, i32 0
  %buf_ptr8 = load [1 x i64]*, [1 x i64]** %buf7, align 8, !tbaa !5
  %buf_ptr_cast9 = bitcast [1 x i64]* %buf_ptr8 to i8*
  %slice_ptr10 = getelementptr inbounds i8, i8* %buf_ptr_cast9, i64 %cstart46
  %slice_val11 = load i8, i8* %slice_ptr10, align 1, !tbaa !7
  %int_val12 = sext i8 %slice_val11 to i64
  %ptrtmp = getelementptr inbounds { i64, i64 }, { i64, i64 }* %tupleptrtmp, i32 0, i32 0
  store i64 %int_val, i64* %ptrtmp, align 8, !tbaa !3
  %ptrtmp13 = getelementptr inbounds { i64, i64 }, { i64, i64 }* %tupleptrtmp, i32 0, i32 1
  store i64 %int_val12, i64* %ptrtmp13, align 8, !tbaa !3
  %tupletmp = load { i64, i64 }, { i64, i64 }* %tupleptrtmp, align 8, !tbaa !3
  %elemtmp = extractvalue { i64, i64 } %tupletmp, 1
  %elemtmp14 = extractvalue { i64, i64 } %tupletmp, 0
  %1 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @int_fmt, i32 0, i32 0), i64 %elemtmp14)
  %2 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @sep_str, i32 0, i32 0))
  %3 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @int_fmt, i32 0, i32 0), i64 %elemtmp)
  %4 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @newline_str, i32 0, i32 0))
  %cstart015 = load i64, i64* %cstart0, align 8, !tbaa !3
  %addtmp16 = add nsw i64 %cstart015, 2
  store i64 %addtmp16, i64* %cstart0, align 8, !tbaa !3
  %i117 = load i64, i64* %i1, align 8, !tbaa !3
  %addtmp18 = add nsw i64 %i117, 1
  store i64 %addtmp18, i64* %i1, align 8, !tbaa !3
  %i119 = load i64, i64* %i1, align 8, !tbaa !3
  %count220 = load i64, i64* %count2, align 8, !tbaa !3
  %lttmp21 = icmp slt i64 %i119, %count220
  br i1 %lttmp21, label %loopbody, label %loopend

loopend:                                          ; preds = %loopbody, %loopcond
  ret void
}

; Function Attrs: argmemonly norecurse nounwind readonly
define void @consumer(%params* noalias %0) #4 {
entry:
  %tupleptrtmp = alloca { i64, i64 }, align 8
  %i1 = alloca i64, align 8
  %cstart4 = alloca i64, align 8
  %cstart3 = alloca i64, align 8
  %cstart0 = alloca i64, align 8
  %count2 = alloca i64, align 8
  store i64 0, i64* %cstart0, align 8, !tbaa !3
  store i64 0, i64* %i1, align 8, !tbaa !3
  store i64 5, i64* %count2, align 8, !tbaa !3
  br label %loopcond

loopcond:                                         ; preds = %entry
  %i11 = load i64, i64* %i1, align 8, !tbaa !3
  %count22 = load i64, i64* %count2, align 8, !tbaa !3
  %lttmp = icmp slt i64 %i11, %count22
  br i1 %lttmp, label %loopbody, label %loopend

loopbody:                                         ; preds = %loopbody, %loopcond
  %cstart03 = load i64, i64* %cstart0, align 8, !tbaa !3
  store i64 %cstart03, i64* %cstart3, align 8, !tbaa !3
  %cstart34 = load i64, i64* %cstart3, align 8, !tbaa !3
  %addtmp = add nsw i64 %cstart34, 1
  store i64 %addtmp, i64* %cstart4, align 8, !tbaa !3
  %cstart35 = load i64, i64* %cstart3, align 8, !tbaa !3
  %buf = getelementptr inbounds %params, %params* %0, i32 0, i32 0
  %buf_ptr = load [1 x i64]*, [1 x i64]** %buf, align 8, !tbaa !5
  %buf_ptr_cast = bitcast [1 x i64]* %buf_ptr to i8*
  %slice_ptr = getelementptr inbounds i8, i8* %buf_ptr_cast, i64 %cstart35
  %slice_val = load i8, i8* %slice_ptr, align 1, !tbaa !7
  %int_val = sext i8 %slice_val to i64
  %cstart46 = load i64, i64* %cstart4, align 8, !tbaa !3
  %buf7 = getelementptr inbounds %params, %params* %0, i32 0, i32 0
  %buf_ptr8 = load [1 x i64]*, [1 x i64]** %buf7, align 8, !tbaa !5
  %buf_ptr_cast9 = bitcast [1 x i64]* %buf_ptr8 to i8*
  %slice_ptr10 = getelementptr inbounds i8, i8* %buf_ptr_cast9, i64 %cstart46
  %slice_val11 = load i8, i8* %slice_ptr10, align 1, !tbaa !7
  %int_val12 = sext i8 %slice_val11 to i64
  %ptrtmp = getelementptr inbounds { i64, i64 }, { i64, i64 }* %tupleptrtmp, i32 0, i32 0
  store i64 %int_val, i64* %ptrtmp, align 8, !tbaa !3
  %ptrtmp13 = getelementptr inbounds { i64, i64 }, { i64, i64 }* %tupleptrtmp, i32 0, i32 1
  store i64 %int_val12, i64* %ptrtmp13, align 8, !tbaa !3
  %tupletmp = load { i64, i64 }, { i64, i64 }* %tupleptrtmp, align 8, !tbaa !3
  %elemtmp = extractvalue { i64, i64 } %tupletmp, 1
  %elemtmp14 = extractvalue { i64, i64 } %tupletmp, 0
  call void asm sideeffect "", "X,~{memory}"(i64 %elemtmp14), !tbaa !9
  call void asm sideeffect "", "X,~{memory}"(i64 %elemtmp), !tbaa !9
  %cstart015 = load i64, i64* %cstart0, align 8, !tbaa !3
  %addtmp16 = add nsw i64 %cstart015, 2
  store i64 %addtmp16, i64* %cstart0, align 8, !tbaa !3
  %i117 = load i64, i64* %i1, align 8, !tbaa !3
  %addtmp18 = add nsw i64 %i117, 1
  store i64 %addtmp18, i64* %i1, align 8, !tbaa !3
  %i119 = load i64, i64* %i1, align 8, !tbaa !3
  %count220 = load i64, i64* %count2, align 8, !tbaa !3
  %lttmp21 = icmp slt i64 %i119, %count220
  br i1 %lttmp21, label %loopbody, label %loopend

loopend:                                          ; preds = %loopbody, %loopcond
  ret void
}

define %params* @create(i64* %0) {
entry:
  %malloccall = tail call i8* @malloc(i32 ptrtoint (%params* getelementptr (%params, %params* null, i32 1) to i32))
  %paramstmp = bitcast i8* %malloccall to %params*
  %buf = getelementptr inbounds %params, %params* %paramstmp, i32 0, i32 0
  %tmpbufp = bitcast i64* %0 to [1 x i64]*
  store [1 x i64]* %tmpbufp, [1 x i64]** %buf, align 8
  ret %params* %paramstmp
}

declare noalias i8* @malloc(i32)

attributes #0 = { argmemonly nofree nosync nounwind willreturn }
attributes #1 = { argmemonly readonly speculatable }
attributes #2 = { argmemonly norecurse nounwind readonly speculatable }
attributes #3 = { readnone }
attributes #4 = { argmemonly norecurse nounwind readonly }

!0 = !{!1, !1, i64 0, i64 0}
!1 = !{!"string", !2}
!2 = !{!"castor_root"}
!3 = !{!4, !4, i64 0, i64 0}
!4 = !{!"runtime", !2}
!5 = !{!6, !6, i64 0, i64 1}
!6 = !{!"db", !2}
!7 = !{!8, !8, i64 0, i64 1}
!8 = !{!"db_int", !6}
!9 = !{!10, !10, i64 0, i64 0}
!10 = !{!"consumer", !2}
